crash.subsystem.filesystem: fix mount command for SLE12-SP3 kernel

luis-henrix opened the pull request:
'mnt_root' access in recent kernels is done through the struct vfsmount
'mnt' field.

Signed-off-by: Luis Henriques <lhenriques@suse.com>

jeffmahoney committed:
initial commit

jeffmahoney committed:
kdump: initial basic target

This module contains the bare minimum to load an x86_64 kernel crash
dump and populate the thread list with thread info and stacks.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
COPYING: add GPLv2 license file

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
.gitignore: ignore *.pyc and *~

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.arch: introduce and use architecture-specific thread handling

This commit defines a CrashArchitecture base class for
architecture-specific code.  We use it for implementing initial x86_64
support for thread handling.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.arch: minor cleanups

- CrashArchitecture needs to be a real class
- Kill %-based string formatting
- Style corrections
- Add call to super().__init__

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.percpu: Add percpu handling

This commit adds support for special handling of percpu variables.  It will
accept symbols that are in the percpu sections, symbols that point to percpu
objects, and passed pointers to percpu objects.  If a symbol or value is
passed that is not a percpu (as determined by whether the offset falls
within the percpu section), a TypeError is raised.

Thanks to Vlastimil Babka for the initial work in extending percpu handling
to values instead of just symbols.

Test cases are also included to test a variety of access methods.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kdump.target: to_thread_alive returns bool

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash: add support for active threads

The kernel data structures in the crash dump tell us which threads were
active when the dump was taken but not the contents of those threads'
registers.  The dump format does contain that information so we need
to access it from the vmcore attributes instead.

This commit also fixes some API issues that remained from a previous
version of libkdumpfile.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands: add initial command interface

Add a simple way to add commands.  Initially, commands have a 'py'
prefix.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

ptesarik committed:
Use kdumpfile module address space constants

Commit ee04e262c9eb70c7272859bb0985d34b55b09aba in libkdumpfile broke
the kdump target, as it changes address space constants to module
constants.

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

jeffmahoney committed:
Merge pull request #1 from ptesarik/master

Use kdumpfile module address space constants

jeffmahoney committed:
crash.cache: add new cache submodule and a skel for vm

We're getting to the point now where some caches would help, so this
sets up the basic infrastructure.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands: add help command

`pyhelp' will now list all of the commands implemented via crash.commands

It will also display the help text for an individual command if specified.

Since the help text is integrated into the gdb.Command stuff as well,
'help <command>' also works, as does 'help user-defined'

Using  after the first help text line makes it so the summary appears
in 'help user-defined' but we get the full text without the truncated
header when getting help for a particular command.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
arch/x86_64: do not fetch rip unless specifically requested

We try to batch register updates as a performance optimization, but
updating 'rip' has a side effect: it will reinitialize the frame cache.

Once upon a time, this caused a segfault with 'info frame' but that turned
out to be a bug in gdb.

So, now we just update rip when it's requested specifically or if
all registers are requested.  Since it's not setting up the thread
anymore, we'll rename the affected routines to fetch_register_*.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

pmladek committed:
sys: Initial implementation of the "sys" command and cache

"sys" command will provide many useful information about the dumped
system.

The initial implementation shows values from struct new_utsname
that is accessible via the global variable init_uts_ns.name.

All the values will get cached because they might be needed
also by other commands.

[jeffm: modified to use delayed initialization]

Signed-off-by: Petr Mladek <pmladek@suse.com>
Signed-off-by: Jeff Mahoney <jeffm@suse.com>

pmladek committed:
sys: Print raw kernel config using "sys config"

Extend the sys command to print the raw kernel config when
called with the "config" parameter.

I am sure that the handling of invalid option might be done
a better way but I do not know how :-)

[jeffm: modified to use delayed initialization]

Signed-off-by: Petr Mladek <pmladek@suse.com>
Signed-off-by: Jeff Mahoney <jeffm@suse.com>

pmladek committed:
sys: get and show uptime

This patch improves the "sys" command to show the uptime.
For this it creates few more caches.

We extend the crash.cache.syscache.config interface to operate
as a dictionary.
+ crash.cache.syscache.config['HZ'] = '250'

We introduce a kernel crash.cache.syscache.kernel interface to
contain calculated or machine-dependent values.
+ crash.cache.syscache.kernel.uptime = "2 days, 13:27:07"
+ crash.cache.syscache.kernel.hz = long(250)

[jeffm: Rewritten to make better use of the Python namespace]

Signed-off-by: Petr Mladek <pmladek@suse.com>
Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands: parser should be optional

For simple commands that don't take any options, we only need the parser
to print the help text.  We can do that in the CrashCommand constructor
instead of forcing every caller to do it.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands: append newline to help text if none exists

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands.help: actually print the help text for a command

We were fetching the help text, but not actually printing it.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands.ps: add -n option to list gdb thread number

It's useful to convert between kernel pids and gdb threads, so using -n
will list the thread number instead of the task struct pointer.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash: add 'thread' command to select a gdb thread by Linux pid

This adds a new 'thread' command that selects the gdb thread for a
Linux pid.  It adds a task cache that the kdump target populates
while creating the tasks to make it fast.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
contrib: add contrib directory for one-off scripts

This commit adds a contrib directory to hold scripts that may not be
ready to be incorporated into the main crash project.

Initially it contains 'mods.py' that loads all modules from
/lib/modules/$UNAME_R/kernel and 'lsbtrfs-cmd.py' that is a demo for
adding a quick command.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

lorddoskias committed:
ps: Explicitly convert gdb.Values to ints

tabraham committed:
add load average to sys command output

[jeffm: rewritten to adhere to style]
Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash: add Makefile and man page

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands.dmesg: properly decode log buf to utf-8

Without doing replace, we get decoding exceptions.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: add @delayed_init, @exporter, and @export decorators

These three decorators codify how code should be structured and exported
when establishing crash-python infrastructure modules.

Classes that require information from the underlying executable or target
must be declared with the @delayed_init decorator.  This delays execution
of the __init__ method until the first attribute that needs initialization
is accessed.  This is critical for implementing unit tests so that
importing modules doesn't attempt to pull in the world as it starts up.

Modules that publish an API, which should be all of them, must locate
all routines in a class declared with the @exporter decorator.  Each
method to be exported to the module-level must be decorated with
@export.  The @exporter decorator instantiates the class as a singleton
and "steals" the class name as the instance name.  It then locates
any methods declared with @export and publishes them in the module
namespace.  Instance, class, and static methods are all supported.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
tests: add very basic testing infrastructure

This patch adds the test-all.sh command that starts up a GDB instance
and runs the test cases within it.  No test cases are added.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
tests: add very basic testing infrastructure

This patch adds the test-all.sh command that starts up a GDB instance
and runs the test cases within it.  No test cases are added.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.util: split util module into util and list

This module implements the container_of and offsetof routines in
addition to some basic type lookup helpers and places them in a new
crash.types.util module.  The list operations are moved into a new
crash.types.list module.  In addtional, cycle detection has been added
by Vlastimil Babka.

Extensive test cases for offsetof and lists are included.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.util: move and improve util module

This module implements the container_of and offsetof routines in
addition to some basic type lookup helpers.

Extensive test cases for offsetof are included.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash: 2to3 conversion

Convert python to code that can be used with Python 3.  All future
changes must be Python 3-safe.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kdump.target: retab

There were some bare tabs lurking in there.  Now it's just spaces as
expected.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
python: replace last remaining %-style strings with .format() style

This commit replaces the last remaining %-style formatted strings with
.format() style strings.  New commits should only use .format() style.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: fix delayed_init and exporter, add real testing

Further development showed that the crash.infra implementation was
incomplete and hacky.  This patch converts delayed_init to a class that
the decorator creates using the original class as a base class.  It
also converts the exporter to use the inspect interface, which does
do what we need and uses bound methods so we don't need to play games
with the wrapper.

Lastly, if @exporter is applied before @delayed_init, it will end up
triggering the delayed initialization during import, which is exactly
what these are intended to avoid.  We'll throw a TypeError in this case
instead of just causing weird bugs.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash: add support for loading modules and debuginfo

This commit adds support for loading kernel modules into the address
space.  It will search a list of directories recursively looking for
modules and load them automatically.  This is a change over legacy
crash but is necessary since we have full stack unwinding and we need
the unwind information from the modules to make that work.

In order to use any of this, we need valid debuginfo.  For self-built
kernels, this will typically still be in the object files but for
kernels we ship, the debuginfo is separate.  We'll search the same
directories for the debuginfo and load it automatically as well.

We also add a simple driver to load everything up.  The easiest
way to use crash is now the crash.sh script.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kdump.target: add debugging output for memory transfers

The GDB python target consumes the exceptions raised by the memory
transfer operations silently, which is normal given how many will
fail during routine operation.  When debugging crash-python itself,
it's helpful to see when those operations fails.  This patch adds
a debugging mode that prints a message when those exceptions are
encountered.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.sh: add valgrind and gdb startup modes for debugging gdb issues

I've discovered several bugs that turned out to be bugs in gdb patches
so having the ability to start it up under gdb or valgrind has proven
very useful.  This patch adds --gdb and --valgrind options to do that.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kdump.target: print updates while loading tasks

Rather than silently churn while loading tasks, print an initial
message and another dot for every 100 tasks so progress is clear.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.arch: add a stack frame filter

gdb allows the use of a Frame Filter, which we can use to ensure
that stack traces don't wander into userspace.  Since the address
limits are per-architecture, the architecture specifies the limit.

This works for x86_64.  Other architectures with more complex layouts
may require additional work.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kdump.target: cleanup API issues

- python-kdumpfile API has changed, so we need updating
- add simple tests
- mark some methods as @staticmethods to keep pylint happy
- make CrashArchitecture a real class and call its __init__

jeffmahoney committed:
crash: add simple 'make lint' target

The 'lint' make target will run pylint on the crash module using
the tests/pylintrc file for configuration.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
tasks: implement `ps' and infrastructure behind it

This provides a fairly complete ps command along with the task
infrastructure that can be used elsewhere.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands.dmesg: add a dmesg command to show the kernel log

It interprets both new structured log format and old flat format logs.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash: python3 compatibility -> long = int

Despite my best efforts with pylint, a few more long() calls snuck in
without checking for python3.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash: python3 compatibility -> except can't have commas

Another pylint slip -- except Exception, e usage is gone in python3.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
lint: add lint3 mode and silence statistics reporting

'make lint3' will now do checks for Python3 portability
'make lint' will do those checks and the original Python2 checks

Also silence the statistics report at the end.  It just ends up
scrolling the useful information off the screen.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.task: PF_EXITING needs to be specified as long(0x4)

The L suffix is gone in Python3, so we need to do long(0x4) to get
it to be a long in Python2.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: document limitation with super().__init__() and delayed_init

There is a major shortcoming with delayed_init inheriting from the base
class behind the scenes: super() will refer to the base class.  As a
result, super().__init__() call from base_class.__init__() will
result in infinite recursion.

It's fixable, but it's also easy enough to work around.  We'll leave it
as a problem to solve on another day.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: properly handle classes with __getattr__

We weren't properly handling delayed init with classes with
__getattr__ so we'd get AttributeError when the attribute didn't
exist after __init__ was called.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands: make more testable

We currently do a few things that makes it difficult to do testing of
commands:
- The command parser exits, which gdb catches and turns into a SystemExit
  exception, but this could mean anything.
- We catch a few exceptions ourselves, so even if we called
  CrashCommand.invoke directly instead of gdb.execute(), we won't actually
  get an normal error condition back -- only the exceptional ones.

This patch moves the command execution into a new invoke_uncaught routine
that doesn't catch exceptions.  The old invoke continues to catch
exceptions, so we don't fall over during normal execution.

We have a new CrashCommandParser that /must/ be used when creating new
commands.  This is so we can raise our own exception that we can
handle in the unit tests.  All existing commands are switched over

We register the command as the last step so we don't trip over the
TypeError later.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.cache.syscache: refactor to improve delayed init and improve testing

This patch refactors some of the syscache code so that it's more easily
testable and also cascades delayed initialization on-demand.

We also add a real compressed kernel config blob so we can properly
test CrashConfigCache as well as test the namespace functionality
with and without a kernel image available.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.exceptions: add crash.exceptions class

The test cases should have been failing since crash.util since this
file was missing.  Unfortunately, the pyc file was still there, so
the tests happily used that.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands.ps: maybe_dead is in task, not ps

In task_state_string, we call self.maybe_dead() but it is not
part of ps - it's part of LinuxTask.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kernel: fix searching for debuginfo files in current directory

We would check to see if the file existed and then ignore it.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kernel: mutate filename before looking in findmap

We store the mutated filename in findmap but then don't use it to
locate the file.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kernel: fix delayed initialization race

Delayed init for crash.kernel loads up the modules symbol -- but we
need __init__ for findmap to load the debuginfo for vmlinux.

The fix is to use __getattr__ directly.  Since there aren't any other
types loaded up, we don't need to use @delayed_init at all.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.blockdev: add block device iteration

This commit adds three modules under crash.types:
- blockdev - iterates over block devices
- classdev - iterates over class devices
- klist - iterates over klists

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.task: use type from value rather than lookup

GDB's type lookup doesn't use the symbol cache so we get different but
identical struct types backing the lookup and the type associated
with a value.  Since the pointers differ, a deep comparison needs to
be made to see if they are identical.  Since task_struct references
a huge number of other structures via pointers, the deep comparison
is very expensive. When loading thousands of tasks, this can be very
time consuming.

We can work around the issue by swapping the lookup type for the value
type when we have a matching one.

We also raise a new exception for bad tasks since we're raising and
formatting it in two places.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.list: detect NULL pointer

This patch means we raise a proper CorruptedListError rather than
a "Value not available" error.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.blockdev: properly handle partitions

The system I used to write and test the blockdev support was a VM without
any partitions.  This patch allows us to return the correct values
for partitions and bdevs in for_each_block_device.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash: use build directory for in-repo execution

The crash.sh and test-all.sh scripts use the repo directory for the
module path.  This can result in modules that existed at one point
but were discarded to be used if the .pyc files still exist.

This patch removes the build/lib/crash directory, rebuilds it with
setup.py build, and uses that for runtime and testing.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.util: pass block and domain to lookup_symbol helpers

In some cases, we need to be able to specify the block and domain
for symbol lookup.  This patch adds the ability to do that.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: add symbol/type callback infrastructure

GDB provides the ability to execute callbacks when objfiles are
loaded.  Rather than blindly delay initialization, we can subscribe
to the callback events and initialize the types and symbols as they
are loaded.

Using the SymbolCallback or TypeCallback classes, the consumer can
have a callback fire once when the symbol or type is available.  A
specific mechanism to assign the result to an attribute in the caller
is also provided.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.subsystem.filesystem: Add filesystem subsystem infrastructure

This patch adds support for the filesystem subsystem.  Initial support
for btrfs and ext3 is provided.  The mount command is implemented.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.cache.syscache: add jiffies_to_msec

We calculate jiffies here and consumers need to convert to wall clock time.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.util: allow passing block to safe_lookup_type

Some contexts require specifying the block for a type lookup.  Pass
it through.  GDB needs to support passing None as a block.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.subsystem.storage: Add storage subsystem infrastructure

This patch adds a set of modules for use with the storage subsystem.

Initially, it provides a bio chain unwinder that allows storage
implementations and file systems to register callbacks that decode
a specific bio.  Basic device mapper support is provided.

This also pulls in crash.types.blockdev and eliminates the old module.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: exporter, create instance on first use

Replacing the class name with an instance was troublesome and
unnecessary.  Now we keep a dictionary of the singleton instances
attached to the module and use a callable wrapper to instantiate
the singleton when required.  This should also translate into needing
to use delayed initialization much less since the instances aren't
created during import - only the wrappers are.  It also means that
the ordering with @exporter and @delayed_init doesn't matter anymore.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: turn into a proper submodule

We'll be adding more things to infra that make sense as separate modules.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash: move Session into session module

We import everything required to setup a session just by importing
anything under crash.*.  That complicates testing.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: add submodule loading helper

Rather than having each portion of crash load itself, have the roots of
each submodule recursively load its constituent modules.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.sh: allow --nofiles mode for testing delayed symbol loading

It's helpful to test the delayed symbol/type error handling by
just not providing a kernel image.  This also shuffles the initialization
in crash.Session so that crash-python is setup for use but without
a kernel image.

Sigend-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: remove delayed_init

Now that we use DelayedLookups everywhere we can remove @delayed_init.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands.dmesg: remove unused imports

dmesg imports delayed_init and exporter but doesn't use them.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.cache.syscache: python 3 compat, import round from builtins

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: replace @exporter with new base class

The @exporter decorator is a hacky way to export to the module namespace.

This commit adds a new metaclass that does the same thing and adds
a base class that uses that metaclass for everything else to inherit from.

Now all that's required is to inherit from the base class and decorate
using @export as before.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.infra: convert simple delayed_init cases to CrashBaseClass

For classes where we just do simple type or symbol lookups, we can
convert them easily in one commit.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.percpu: convert to use CrashBaseClass lookups

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.task: convert to use CrashBaseClass lookups

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kernel: silence noisy debuginfo messages

The debuginfo loading messages are noisy.  We can use a symtab lookup
to see if we need the debuginfo before loading it and report an error
if we can't.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.blockdev: convert to use CrashBaseClass lookups

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

ptesarik committed:
crash.kdump.target: Reflect recent changes to kdumpfile module API

SysErrException was renamed to OSErrorException.
kdumpfile.vtop_init() is gone; instead, set the 'addrxlat.ostype'
attribute to the desired OS type (always 'linux' for now).

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

ptesarik committed:
crash.kdump.target: Re-introduce symbol callback

The current code works just fine if correct information can be found in
the dump file itself (using VMCOREINFO). However, if this information is
not available, address translation may fail.

The new code still prefers data from VMCOREINFO, because that is more
likely to match the dump file, but if the default method fails, it
consoluts GDB symbol table.

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

jeffmahoney committed:
Merge pull request #19 from ptesarik/master

Adapt to libkdumpfile API changes

ptesarik committed:
Fix a braino in BadTaskError.__init__()

Instantiation of BadTaskError fails because of this code (most likely
written too late at night).

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

jeffmahoney committed:
Merge pull request #20 from ptesarik/master

Fix a braino in BadTaskError.__init__()

jeffmahoney committed:
crash.cache.syscache: convert to use delayed lookups

This converts syscache to use delayed lookups and updates the test cases.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.sh: add help text for debugging options

I added the --gdb, --valgrind, and --nofiles options but didn't document
them.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.types.percpu: don't use section to determine percpu status

jeffmahoney committed:
gdb-update: Fix of MinimalSymbol.value() use

We need to use .address instead of just the value.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands: convert to use CrashBaseClass

This allows commands to use delayed lookups.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.commands.dmesg: convert dmesg to use delayed lookups

This converts dmesg to use delayed lookups instead of loading the symbols
at each use.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.kdump.target: fix typo that ignored threads

Commit ee9b80b6 (crash: add support for active threads) added iteration
over the threads in thread groups but still iterated over the task
list.  As a result, we skipped those threads.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

jeffmahoney committed:
crash.subsystem.storage: don't raise an exception on undecoded bio

There's no need to abort iteration over bios when one can't be
decoded.  Just report it as undecoded and continue.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

ptesarik committed:
CrashConfigCache: Use the gzip header

From zlib documentation of inflateInit2():

> windowBits can also be greater than 15 for optional gzip decoding.  Add
> 32 to windowBits to enable zlib and gzip decoding with automatic header
> detection, or add 16 to decode only the gzip format (the zlib format will
> return a Z_DATA_ERROR).  If a gzip stream is being decoded, strm->adler is
> a crc32 instead of an adler32.

Since kernel_config_data is always compressed with gzip, let's use 16.

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

ptesarik committed:
Fix a braino in crash.util.safe_lookup_type()

The block argument should not be always None...

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

ptesarik committed:
vtop: Initial implementation of the "vtop" command

Add a crash module for generic address translation and use it for a
"vtop" command. This is a minimal implementation which shows only
the resulting physical address, not the translation steps.

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

ptesarik committed:
vtop: Print out details on page table walk

Use the low-level API to show detailed information about the page
table walk.

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

ptesarik committed:
vtop: Reuse libkdumpfile address translation if possible

The address translation system is currently created from scratch
using built-in heuristics and data from gdb. However, there is
already a translation system when working with a libkdumpfile
target.

BTW this is the only way to make address translation work on some
dump files, e.g. Xen xc_core ELF dump files contain the mapping
between physical and machine addresses in the non-data portion of
the file. In other words, it is not mapped to any virtual address,
so the only way to reach it is through direct access to the dump
file.

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

ptesarik committed:
vtop: Include physical address in Xen non-auto-translated domains

For non-auto-translated domains, the value stored in the pagetable
is the machine physical address. Let's also include the kernel
physical address in the output.

Signed-off-by: Petr Tesarik <ptesarik@suse.com>

jeffmahoney committed:
Merge pull request #21 from ptesarik/master

Implement vtop command

jeffmahoney commented on crash/subsystem/filesystem/mount.py:
```diff
@@ -126,6 +126,15 @@ def mount_super(mnt):
         return sb
 
     @export
+    @staticmethod
```
This should be handled as a callback where we have two versions, one that returns mnt['mnt_root'] and the other that returns mnt['mnt']['mnt_root'].  See crash.types.task.pick_get_rss for an example.

luis-henrix replied:
Hmm, ok that does look a bit better.  I'll send you another pull request later, thanks

Commit:
crash.subsystem.filesystem: fix mount command for SLE12-SP3 kernel

With more recent kernels for_each_mount() returns a 'struct mount' instead
of 'struct vfsmount'.  Added mnt_root() so that it can transparently
access 'mnt_root' field from the right place.

Signed-off-by: Luis Henriques <lhenriques@suse.com>

luis-henrix commented:
I've re-pushed the mount-fix branch again.  It actually fixes the issue in a different way, as I realised that the for_each_mount() actually returns 'struct mount' in 12SP3, while it was returning 'struct vfsmount' in older kernels.

luis-henrix commented:
Ok, I've pushed v3 into branch mount-fix-v3.  It's a bit more generic fix, which allows to also fix the '-f' option.
